### PR TITLE
migration: Add case about vhostvdpa

### PR DESCRIPTION
**Pull request by cliping**
XXX-300493 - [vhostvdpa-blk] Migrate vm with disk of vhostvdpa backend

### CODE REVIEW

**Comment by cliping** [comment]
Test result:

 (1/2) type_specific.io-github-autotest-libvirt.migration_with_copy_storage.migration_with_vhostvdpa.copy_storage_all.p2p: PASS (410.15 s)
 (2/2) type_specific.io-github-autotest-libvirt.migration_with_copy_storage.migration_with_vhostvdpa.copy_storage_all.non_p2p: PASS (411.92 s)

File: libvirt/tests/cfg/migration_with_copy_storage/migration_with_vhostvdpa.cfg [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    # SSH connection time out
+    ssh_timeout = 60
+    # Local URI
+    virsh_migrate_connect_uri = "qemu:///system"
+    image_convert = "no"
+    migrate_desturi_port = "16509"
+    migrate_desturi_type = "tcp"
+    virsh_migrate_desturi = "qemu+tcp://${migrate_dest_host}/system"
+    server_ip = "${migrate_dest_host}"
+    server_user = "root"
```
Comment (dzhengfy) [inline libvirt/tests/cfg/migration_with_copy_storage/migration_with_vhostvdpa.cfg:18]: Please read it from base.cfg

File: libvirt/tests/cfg/migration_with_copy_storage/migration_with_vhostvdpa.cfg [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    server_pwd = "${migrate_dest_pwd}"
+    client_ip = "${migrate_source_host}"
+    client_user = "root"
```
Comment (dzhengfy) [inline libvirt/tests/cfg/migration_with_copy_storage/migration_with_vhostvdpa.cfg:21]: Please read it from base.cfg

File: libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+        test.log.info("Setup steps.")
+        migration_obj.setup_connection()
+
+        setup_vhost_vdpa_disk(params, run_on_target)
+        # Set shared memory
+        vm_xml.VMXML.set_memoryBacking_tag(vm_name, access_mode="shared", hpgs=False)
+        vmxml = vm_xml.VMXML.new_from_dumpxml(vm_name)
+        vmxml.add_device(libvirt_vmxml.create_vm_device_by_type("disk", disk_dict))
+        vmxml.sync()
```
Comment (dzhengfy) [inline libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py:71]: I did not see any vm xml backup. But it seems we need, right?

**Review by dzhengfy — COMMENTED** [review]
Others LGTM

File: libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+from virttest.libvirt_xml import vm_xml
+from virttest.utils_libvirt import libvirt_disk
+from virttest.utils_libvirt import libvirt_vmxml
+
+from provider.migration import base_steps
+
+
+def setup_vhost_vdpa_disk(params, run_on_target):
+    cmd = "modprobe vhost-vdpa"
+    process.run(cmd, shell=True, verbose=True)
```
Comment (Yingshun) [inline libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py:25]: utils_vdpa.VDPASimulatorTest() can setup/cleanup the local machine env, would you like to reuse it?

File: libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    remote.run_remote_cmd(cmd, params, run_on_target)
+
+    cmd = "vdpa dev add mgmtdev vdpasim_blk name blk0"
+    process.run(cmd, shell=True, verbose=True)
+    remote.run_remote_cmd(cmd, params, run_on_target)
+
+
+def cleanup_vhost_vdpa_disk(params, run_on_target):
+    cmd = "modprobe -r vdpa-sim-blk"
+    process.run(cmd, shell=True, verbose=True)
```
Comment (Yingshun) [inline libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py:39]: Same as above

File: libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+from virttest import remote
+
+from virttest.libvirt_xml import vm_xml
+from virttest.utils_libvirt import libvirt_disk
+from virttest.utils_libvirt import libvirt_vmxml
+
+from provider.migration import base_steps
+
+
+def setup_vhost_vdpa_disk(params, run_on_target):
```
Comment (Yingshun) [inline libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py:23]: plz update docstring, thx!

File: libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    cmd = "modprobe vdpa-sim-blk"
+    process.run(cmd, shell=True, verbose=True)
+    remote.run_remote_cmd(cmd, params, run_on_target)
+
+    cmd = "vdpa dev add mgmtdev vdpasim_blk name blk0"
+    process.run(cmd, shell=True, verbose=True)
+    remote.run_remote_cmd(cmd, params, run_on_target)
+
+
+def cleanup_vhost_vdpa_disk(params, run_on_target):
```
Comment (Yingshun) [inline libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py:37]: docstring plz

File: libvirt/tests/cfg/migration_with_copy_storage/migration_with_vhostvdpa.cfg [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    start_vm = "yes"
+    setup_local_nfs = "yes"
+    storage_type = "nfs"
+    mnt_path_name = ${nfs_mount_dir}
+    virsh_migrate_extra = "--migrate-disks vdb"
+    status_error = "no"
+    check_disk_after_mig = "yes"
+    vhost_vdpa_disk = "/dev/vhost-vdpa-0"
+    disk_dict = {"device": "disk", "type_name": "vhostvdpa", "driver": {"name": "qemu", "type": "raw", "cache": "none", "io": "threads", "copy_on_read": "on", "discard": "unmap", "detect_zeroes": "on"}, "target": {"dev": "vdb", "bus": "virtio"}, "source": {"attrs": {"dev": "${vhost_vdpa_disk}"}}}
+    func_supported_since_libvirt_ver = (10, 0, 0)
```
Comment (Yingshun) [inline libvirt/tests/cfg/migration_with_copy_storage/migration_with_vhostvdpa.cfg:32]: Is this case supported on all archs?

File: libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py (hunk 6 above)
Comment (chunfuwen) [reply libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py:23]: setup_vhost_vdpa_disk-->setup_vhost_vdpa_simulator_disk ?

File: libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py (hunk 7 above)
Comment (chunfuwen) [reply libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py:37]: cleanup_vhost_vdpa_disk --> cleanup_vhost_vdpa_simulator_disk

File: libvirt/tests/cfg/migration_with_copy_storage/migration_with_vhostvdpa.cfg (hunk 1 above)
Comment (cliping) [reply libvirt/tests/cfg/migration_with_copy_storage/migration_with_vhostvdpa.cfg]: Updated, thanks.

File: libvirt/tests/cfg/migration_with_copy_storage/migration_with_vhostvdpa.cfg (hunk 2 above)
Comment (cliping) [reply libvirt/tests/cfg/migration_with_copy_storage/migration_with_vhostvdpa.cfg]: Updated, thanks.

File: libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py (hunk 3 above)
Comment (cliping) [reply libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py:83]: In class MigrationBase(), line 57 - 59, backup vm xml. Thanks.

File: libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py (hunk 4 above)
Comment (cliping) [reply libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py]: Updated, thanks.

File: libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py (hunk 5 above)
Comment (cliping) [reply libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py]: Updated, thanks.

File: libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py (hunk 6 above)
Comment (cliping) [reply libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py]: Updated, thanks.

File: libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py (hunk 7 above)
Comment (cliping) [reply libvirt/tests/src/migration_with_copy_storage/migration_with_vhostvdpa.py]: Updated, thanks.

File: libvirt/tests/cfg/migration_with_copy_storage/migration_with_vhostvdpa.cfg (hunk 8 above)
Comment (cliping) [reply libvirt/tests/cfg/migration_with_copy_storage/migration_with_vhostvdpa.cfg:32]: Add 'only x86_64' in cfg, thanks.

**Review by dzhengfy — APPROVED** [review]

**Review by Yingshun — APPROVED** [review]

**Pull request merged**